gcc/rust/resolve/rust-forever-stack.h

```diff
@@ -0,0 +1,531 @@
+resolution passes. We will revisit the ribs we created in these passes, and we
+won't need to allocate or create new ones: because they will still be present in
+the stack, we will simply move our cursor to these ribs. In this case, there is
+nothing to do, since there are no uses of our definitions, as the Rust code we
+are name-resolving is not really interesting. You'll also note that our
+`TopLevel` pass did not resolve a whole lot: all it did was create new ribs, and
+empty ones at that. The `Early` pass will not go further, since our code does
+not contain any imports, macro definitions or macro invocations. You can look at
+this pass's documentation for more details on this resolution process.
+
+**/
+template <Namespace N> class ForeverStack
+{
+public:
+  ForeverStack ()
+    : root (Node (Rib (Rib::Kind::Normal))), cursor_reference (root)
+  {
+    rust_assert (root.is_root ());
+    rust_assert (root.is_leaf ());
+  }
+
+  /**
+   * Add a new Rib to the stack. If the Rib already exists, nothing is pushed
+   * and the stack's cursor is simply moved to this existing Rib.
+   *
+   * @param rib The Rib to push
+   * @param id The NodeId of the node for which the Rib was created. For
+   *        example, if a Rib is created because a lexical scope is entered,
+   *        then `id` is that `BlockExpr`'s NodeId.
+   * @param path An optional path if the Rib was created due to a "named"
+   *        lexical scope, like a module's.
+   */
+  void push (Rib rib, NodeId id, tl::optional<Identifier> path = {});
+
+  /**
+   * Pop the innermost Rib from the stack
+   */
+  void pop ();
+
+  /**
+   * Insert a new definition in the innermost `Rib` in this stack
+   *
+   * @param name The name of the definition
+   * @param id Its NodeId
+   *
+   * @return `AlreadyInserted` if that node was already present in the Rib, the
+   *         node's `NodeId` otherwise.
+   *
+   * @aborts if there are no `Rib`s inserted in the current map, this function
+   *         aborts the program.
+   */
+  tl::expected<NodeId, AlreadyInserted> insert (Identifier name, NodeId id);
+
+  /* Access the innermost `Rib` in this map */
+  Rib &peek ();
+  const Rib &peek () const;
+
+  /**
+   * Reverse iter on all ribs from the innermost one to the outermost one,
+   * trying to find a name. This is the default algorithm.
+   * This function gets specialized based on the Rib::Kind
+   * this way, we ensure a proper resolution algorithm at the type level
+   *
+   * @param name Name of the identifier to locate in this scope or an outermore
+   *        scope
+   *
+   * @return a valid option with the NodeId if the identifier is present in the
+   *         current map, an empty one otherwise.
+   */
+  inline tl::optional<NodeId> get (const Identifier &name);
+
+  // TODO: Should we dump to a file instead, like the other dumps? Or just print
+  // the string to a file in the SessionManager?
+  std::string debug_str ();
```

Comment:
This is the right thing to do because you can do assertions and call rust_debug in dev. Writing to file should be something where we change rust_debug to be a logger of some kind.